### PR TITLE
feat: support for optional dependencies

### PR DESCRIPTION
**Pull request by dsherret**

### CODE REVIEW

File: src/resolution/graph.rs [hunk 1]
```diff
@@ -623,17 +623,43 @@ impl Graph {
...
+          dist: version_info.dist.clone(),
           dependencies,
         },
       );
     }
 
+    // only bother doing this if an optional dependency was found
+    if had_optional {
+      for package in packages.values_mut() {
+        package.optional = !required_packages.contains(&package.pkg_id.nv);
```
Comment (dsherret) [inline src/resolution/graph.rs:659]: I was just writing tests for this and realized this doesn't work for required dependencies of optional dependencies.

**Review by bartlomieju — APPROVED** [review]
LGTM, with a minor nitpick

File: src/registry.rs [hunk 2]
```diff
@@ -334,6 +340,26 @@ impl TestNpmRegistryApi {
       .insert(package_to.0.to_string(), package_to.1.to_string());
   }
 
+  pub fn add_optional_dependency(
+    &self,
+    package_from: (&str, &str),
+    package_to: (&str, &str),
```
[the remark this answers is not in the document]
Comment (dsherret) [reply src/registry.rs]: I improved the signatures (this is test code so not putting a lot of effort into documenting it when someone can look at its usage for examples).

**Review by bartlomieju — APPROVED** [review]
LGTM again :)

**Pull request merged**